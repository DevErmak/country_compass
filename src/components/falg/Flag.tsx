import './flag.css';

type Props = {
  flags: string;
  flagsAlt: string;
  coatOfArms: string;
};

export default function Flag({ flags, flagsAlt, coatOfArms }: Props) {
  return (
    <div className="container-flag">
      <div className="flag">Flag</div>
      <img className="flag-img" src={flags} alt={flagsAlt} />
      <div className="coat-of-arms">Coat of arms</div>
      <img className="coat-of-arms-img" src={coatOfArms} alt={''} />
    </div>
  );
}
