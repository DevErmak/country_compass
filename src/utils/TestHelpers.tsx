/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import ReactDOM from 'react-dom';
import { configureStore, combineReducers } from '@reduxjs/toolkit';
import type { Store } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';
import { FormProvider, useForm } from 'react-hook-form';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import infoCountriesReducer from '../store/country/infoCountrySlice';
import infoUserReducer from '../store/user/infoUserSlice';
import createSagaMiddleware from 'redux-saga';
export const saga = createSagaMiddleware();

const rootReducer = combineReducers({
  infoCountries: infoCountriesReducer,
  infoUser: infoUserReducer,
});

interface IFormWrapperProps {
  children: React.ReactNode;
  defaultValues: Record<string, any>;
}

interface IInitialEntry {
  pathname: string;
  hash?: string;
  key?: string;
  search?: string;
  state?: any;
}

const FormWrapper: React.FC<IFormWrapperProps> = ({ defaultValues, children }) => {
  const formMethods = useForm<typeof defaultValues>({
    defaultValues,
  });

  return <FormProvider<typeof defaultValues> {...formMethods}>{children}</FormProvider>;
};

type RenderComponentWithStoreArgs = {
  Component: React.FC<any>;
  children?: React.ReactNode;
  defaultFormValues?: Record<string, any>;
  memoryRouterInitialEntries?: IInitialEntry[];
  preloadedState?: Record<string, any>;
  props?: Record<string, any>;
};

export const renderComponentWithStore = ({
  Component,
  props,
  preloadedState,
  defaultFormValues,
  memoryRouterInitialEntries = [{ pathname: '/' }],
  children,
}: RenderComponentWithStoreArgs): {
  container: HTMLElement;
  mockStore: Store;
  rerender: (rerenderProps: Record<string, any>) => void;
} => {
  const mockStore = configureStore({
    reducer: rootReducer,
    middleware: [saga],
    preloadedState,
  });

  const renderComponent = (renderProps?: Record<string, any>) => (
    <Provider store={mockStore}>
      <MemoryRouter initialEntries={memoryRouterInitialEntries}>
        {defaultFormValues ? (
          <FormWrapper defaultValues={defaultFormValues}>
            <Component {...renderProps}>{children}</Component>
          </FormWrapper>
        ) : (
          <Component {...renderProps}>{children}</Component>
        )}
        <div id="portal"></div>
      </MemoryRouter>
    </Provider>
  );

  const { container, rerender: reactRerender } = render(renderComponent(props));

  const rerender = (rerenderProps: Record<string, any>) =>
    reactRerender(renderComponent(rerenderProps));

  return { container, mockStore, rerender };
};

type RenderComponentArgs = {
  Component: React.FC<any>;
  children?: React.ReactNode;
  defaultFormValues?: Record<string, any>;
  memoryRouterInitialEntries?: IInitialEntry[];
  props?: Record<string, any>;
};

export const renderComponent = ({
  Component,
  props,
  defaultFormValues,
  memoryRouterInitialEntries = [{ pathname: '/' }],
  children,
}: RenderComponentArgs): { container: HTMLElement } => {
  const { container } = render(
    <MemoryRouter initialEntries={memoryRouterInitialEntries}>
      {defaultFormValues ? (
        <FormWrapper defaultValues={defaultFormValues}>
          <Component {...props}>{children}</Component>
        </FormWrapper>
      ) : (
        <Component {...props}>{children}</Component>
      )}
      <div id="portal"></div>
    </MemoryRouter>,
  );

  return { container };
};
