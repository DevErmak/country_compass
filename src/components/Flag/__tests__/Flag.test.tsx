import Flag from '../Flag';
import { renderComponent } from '../../../utils/TestHelpers';

it('renders correctly Flag', () => {
  const { container } = renderComponent({
    Component: Flag,
    props: { flags: '', flagsAlt: '', coatOfArms: '' },
  });
  expect(container).toMatchSnapshot();
});
