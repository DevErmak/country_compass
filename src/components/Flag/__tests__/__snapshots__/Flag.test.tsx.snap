// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly Flag 1`] = `
<div>
  <div
    class="container-flag"
  >
    <div
      class="flag"
    >
      Flag
    </div>
    <img
      alt=""
      class="flag-img"
      src=""
    />
    <div
      class="coat-of-arms"
    >
      Coat of arms
    </div>
    <img
      alt=""
      class="coat-of-arms-img"
      src=""
    />
  </div>
  <div
    id="portal"
  />
</div>
`;
